src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -295,17 +365,22 @@ public static bool IsMessageHandler(Type type)
                 .Any(genericTypeDef => genericTypeDef == IHandleMessagesType);
         }
 
+        readonly AsyncAutoResetEvent transitionBetweenRateLimitModes = new AsyncAutoResetEvent();
```

Comment:
I think we can delete all the borrowed code by using a TaskCompletionSource similar to what I have done here

https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs#L42

See

https://github.com/Azure/azure-sdk-for-net/pull/6577